crates/server/src/store/in_memory/authority.rs

```diff
@@ -722,6 +767,121 @@ impl InnerInMemory {
+            self.records.remove(&key);
+        }
+
+        // now go through and generate the nsec3 records
+        let ttl = self.minimum_ttl(origin);
+        let serial = self.serial(origin);
+        // FIXME: Should be configurable
+        let opt_out = false;
+
+        let mut records: Vec<Record> = vec![];
+
+        // Store the record types of each domain name so we can generate NSEC3 records for each
+        // domain name.
+        let mut record_types = HashMap::<LowerName, HashSet<RecordType>>::new();
+
+        for key in self.records.keys() {
+            // Store the type of the current record under its domain name
+            match record_types.entry(key.name.clone()) {
+                Entry::Occupied(mut entry) => {
+                    entry.get_mut().insert(key.record_type);
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(HashSet::from([key.record_type]));
+                }
+            }
+
+            // For every domain name between the current name and the origin, add it to
+            // `record_types` without any record types. This covers all the empty non-terminals
+            // that must have an NSEC3 record as well.
+            let mut name = key.name.base_name();
+
+            for _ in origin.num_labels()..name.num_labels() {
+                if let Entry::Vacant(entry) = record_types.entry(name.clone()) {
+                    entry.insert(HashSet::new());
+                }
+                name = name.base_name();
+            }
+        }
+
+        // Compute the hash of all the names.
+        let mut record_types = record_types
+            .into_iter()
+            .map(|(name, type_bit_maps)| {
+                let hashed_name = hash_alg.hash(salt, name.borrow(), iterations).unwrap();
+                (hashed_name, type_bit_maps)
+            })
+            .collect::<Vec<_>>();
+        // Sort by hash.
+        record_types.sort_by(|(a, _), (b, _)| a.as_ref().cmp(b.as_ref()));
+
+        // Generate an NSEC3 record for every name
+        for (i, (hashed_name, type_bit_maps)) in record_types.iter().enumerate() {
+            // Get the next hashed name following the hash order.
+            let next_index = (i + 1) % record_types.len();
+            let next_hashed_name = record_types[next_index].0.as_ref().to_vec();
+
+            let mut type_bit_maps: Vec<RecordType> = type_bit_maps.iter().copied().collect();
+            type_bit_maps.push(RecordType::NSEC3);
+
+            let rdata = NSEC3::new(
+                hash_alg,
+                opt_out,
+                iterations,
+                salt.to_vec(),
+                next_hashed_name,
+                type_bit_maps,
+            );
+
+            let name = Name::new()
+                .append_label(data_encoding::BASE32_DNSSEC.encode(hashed_name.as_ref()))
+                .unwrap()
+                .append_name(origin.borrow())
+                .unwrap();
```

Comment:
I think these unwrap calls are unsafe - a very long zone name could result in the NSEC hashes causing the total record length to be > 255 bytes, which will cause a panic here.